File: rainbow-lib/src/main/java/com/juankysoriano/rainbow/core/event/FingerPositionSmoother.java
package com.juankysoriano.rainbow.core.event;

import com.juankysoriano.rainbow.core.math.RainbowMath;

class FingerPositionSmoother {
    private static final float SPRING = 0.1f;
    private static final float DAMP = 0.6f;
    private float x;
    private float y;
    private float xOld;
    private float yOld;
    private float xVel;
    private float yVel;

    public float getX() {
        return x;
    }

    public float getY() {
        return y;
    }

    float getOldX() {
        return xOld;
    }

    float getOldY() {
        return yOld;
    }

    void moveTo(float x, float y) {
        backupPosition();
        updateVelocity(x, y);
        updatePosition();
    }

    private void backupPosition() {
        xOld = x;
        yOld = y;
    }

    private void updateVelocity(float newX, float newY) {
        float xDistance = (newX - x) * SPRING;
        float yDistance = (newY - y) * SPRING;

        xVel = (xVel + xDistance) * DAMP;
        yVel = (yVel + yDistance) * DAMP;
    }

    private void updatePosition() {
        x += xVel;
        y += yVel;
    }

    void resetTo(float x, float y) {
        this.x = x;
        this.y = y;
        this.xOld = x;
        this.yOld = y;
    }

    float getFingerVelocity() {
        return RainbowMath.dist(0, 0, xVel, yVel);
    }
}

File: rainbow-lib/src/main/java/com/juankysoriano/rainbow/core/touch/FingerPositionSmoother.kt
package com.juankysoriano.rainbow.core.touch

import com.juankysoriano.rainbow.core.math.RainbowMath

private const val SPRING = 0.1f
private const val DAMP = 0.6f
internal class FingerPositionSmoother {
    var x: Float = 0f
        private set
    var y: Float = 0f
        private set
    var oldX: Float = 0f
        private set
    var oldY: Float = 0f
        private set
    private var xVel: Float = 0f
    private var yVel: Float = 0f

    val fingerVelocity: Float
        get() = RainbowMath.dist(0f, 0f, xVel, yVel)

    fun moveTo(x: Float, y: Float) {
        oldX = this.x
        oldY = this.y
        updateVelocity(x, y)
        updatePosition()
    }

    private fun updateVelocity(newX: Float, newY: Float) {
        val xDistance = (newX - x) * SPRING
        val yDistance = (newY - y) * SPRING

        xVel = (xVel + xDistance) * DAMP
        yVel = (yVel + yDistance) * DAMP
    }

    private fun updatePosition() {
        x += xVel
        y += yVel
    }

    fun resetTo(x: Float, y: Float) {
        this.x = x
        this.y = y
        this.oldX = x
        this.oldY = y
    }
}
